### PR TITLE
Add repeat-until functionality to subcircuits

### PR DESCRIPTION
**Pull request by daxfohl**
Adds repeat-until functionality to subcircuits. It's only valid on subcircuits that do not use repetition ids, as otherwise there wouldn't be any mutating measurement value to repeat until.

@95-martin-orion no urgency here, and if we want a different approach for this I'm fine with closing it.

### CODE REVIEW

File: cirq-core/cirq/circuits/circuit_operation.py [hunk 1]
```diff
@@ -94,6 +95,8 @@ class CircuitOperation(ops.Operation):
             will have its path prepended with the repetition id for each
             repetition. When False, this will not happen and the measurement
             key will be repeated.
+        repeat_until: A condition that will be tested after each iteration of
+            the circuit.
```
Comment (95-martin-orion) [inline cirq-core/cirq/circuits/circuit_operation.py:99]: This should clarify that the subcircuit repeats until this condition is true, and include a note that it is incompatible with repetitions.

File: cirq-core/cirq/circuits/circuit_operation.py [hunk 2]
```diff
@@ -280,8 +310,16 @@ def _decompose_(self) -> Iterator['cirq.Operation']:
...
     def _act_on_(self, args: 'cirq.OperationTarget') -> bool:
-        for op in self._decompose_():
-            protocols.act_on(op, args)
+        if self.repeat_until:
+            circuit = self._mapped_single_loop()
+            while True:
+                for op in circuit.all_operations():
+                    protocols.act_on(op, args)
+                if self.repeat_until.resolve(args.classical_data):
+                    break
```
Comment (95-martin-orion) [inline cirq-core/cirq/circuits/circuit_operation.py:319]: It seems like this is equivalent to C++ `do ... while`, i.e. the loop will always execute at least once regardless of the condition. This is a reasonable choice (users shouldn't need to "initialize" a measurement key to use it here), but given that `while` is the more common construct we should document this in the `repeat_until` docstring.

File: cirq-core/cirq/circuits/circuit_operation_test.py [hunk 3]
```diff
@@ -928,4 +946,64 @@ def test_simulate_no_repetition_ids_inner(sim):
...
+                cirq.measure(q, key='m'),
+            ),
+            use_repetition_ids=False,
+            repeat_until=cirq.KeyCondition(key),
+        ),
+    )
+    result = sim.run(c)
+    assert result.records['m'][0][-1] == (1,)
+    for i in range(len(result.records['m'][0]) - 1):
+        assert result.records['m'][0][i] == (0,)
```
Comment (95-martin-orion) [inline cirq-core/cirq/circuits/circuit_operation_test.py:966]: [No change needed] Very cool to see a postselection-esque test working in Cirq.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Very cool. I think I'd like @mpharrigan's input on this as a potential user of this feature (e.g. would this work for measurement-based QC?), but overall I'm quite happy with how well `Condition`s work here.

File: cirq-core/cirq/circuits/circuit_operation.py [hunk 4]
```diff
@@ -223,6 +239,8 @@ def _control_keys_(self) -> AbstractSet['cirq.MeasurementKey']:
                 if not protocols.control_keys(self.circuit)
                 else protocols.control_keys(self.mapped_circuit())
             )
+            if self.repeat_until is not None:
```
Comment (daxfohl) [inline cirq-core/cirq/circuits/circuit_operation.py:242]: This should be removed. We define control keys as things that must be defined prior to the operation. That would be true for a "do-while", but a "repeat-until" does not depend on anything prior to the subcircuit itself. It depends only on the measurement occurring within the subcircuit, which is checked in `_post_init_`.

File: cirq-core/cirq/circuits/circuit_operation_test.py [hunk 5]
```diff
@@ -928,4 +946,64 @@ def test_simulate_no_repetition_ids_inner(sim):
...
 
 
+@pytest.mark.parametrize('sim', [cirq.Simulator(), cirq.DensityMatrixSimulator()])
+def test_repeat_until(sim):
+    q = cirq.LineQubit(0)
+    key = cirq.MeasurementKey('m')
+    c = cirq.Circuit(
+        cirq.CircuitOperation(
+            cirq.FrozenCircuit(
+                cirq.X(q) ** 0.2,
```
Comment (daxfohl) [inline cirq-core/cirq/circuits/circuit_operation_test.py:956]: Just use regular `X` and make sure that it exits after two loops (this should allow Clifford simulation too).

File: cirq-core/cirq/circuits/circuit_operation.py (hunk 2 above)
Comment (daxfohl) [reply cirq-core/cirq/circuits/circuit_operation.py:319]: My first language was Pascal :).

There's probably opportunity for having both this and a plain `while` (that checks the condition prior to the first run). Note a `while` has slightly different semantics in that it requires the first measurement to occur before the circuit, so thus requires to be added to control keys. (I also added repeat-until's controls to the control keys here, but just noticed I shouldn't have, so will remove it in the next commit).

**Comment by daxfohl** [comment]
Should we say this closes #461?

**Comment by 95-martin-orion** [comment]
> Should we say this closes #461?

It's good to mention this on that issue, but I don't think it fully closes it. There are simulator-based optimizations for explicit post-selection (e.g. forcing measurements to a specific result) that this doesn't support.

**Comment by daxfohl** [comment]
Also tagging #3234

File: cirq-core/cirq/circuits/circuit_operation.py (hunk 4 above)
Comment (95-martin-orion) [reply cirq-core/cirq/circuits/circuit_operation.py]: This prevents something like:
```
measure q0, key=k1
circuit operation:
  measure q1, key=k2
  repeat until k1 == k2
```
Is there a low-cost way to allow this behavior? I don't have a specific use case in mind, but this seems like the sort of edge case that a user would be surprised to find doesn't work.

**Comment by daxfohl** [comment]
Would a sympy condition not work? (not at home now)

File: cirq-core/cirq/circuits/circuit_operation.py (hunk 4 above)
Comment (daxfohl) [reply cirq-core/cirq/circuits/circuit_operation.py]: I think this should work already with sympy. Will write a test when I get home.

Comment (95-martin-orion) [reply cirq-core/cirq/circuits/circuit_operation.py]: It works in simulation, but `k1` won't appear in its control keys, possibly leading to misbehavior during construction.

Comment (daxfohl) [reply cirq-core/cirq/circuits/circuit_operation.py]: I see, yeah and commutation assumptions would be broken. Will fix this up when I get home.

Comment (daxfohl) [reply cirq-core/cirq/circuits/circuit_operation.py:246]: Fixed (still slightly different from the original implementation--only *unbound* keys (k1 in your example) are added to the control_keys response), and added test. Good catch!

**Review by 95-martin-orion — APPROVED** [review]
Opted to merge #5033 first since it's smaller - this will need conflict resolution after that finishes.

**Comment by daxfohl** [comment]
@95-martin-orion no conflicts

**Pull request merged**